Intern Property instances so reference equality is guaranteed

magneticflux- opened the pull request:
Closes #19

magneticflux- committed:
Add caching of BooleanProperty instances

Uses WeakReferences, so properties can be GC'd if necessary

magneticflux- committed:
Major refactor to use WeakHashMap and MapKeyGetter instead of FinalizableWeakReferenceToCachedProperty

This allows mixins into Direction and Integer properties as well

magneticflux- committed:
Remove unused parameters from mixins

magneticflux- committed:
Document WeakPropertyCache

magneticflux- committed:
Remove dev logging

2No2Name commented:
I think the asymmetry of Object::equals is problematic. Is there any guarantee that WeakHashMap handles this the way you intend on all platforms? Could there be a platform where equals is called on the stored value leading to this breaking?
Would it be possible to instead use the Property as the key and the value being the key wrapped in a WeakReference? Then getting the value for any equal Property will return the original object, and you still keep the WeakReference behavior of the WeakHashMap.

magneticflux- commented:
I don't know of any implementations that _don't_ call equals on the new object, but this change doesn't seem to have any downsides and it's always good to be cautious, especially with undefined behavior!

magneticflux- committed:
Use a WeakReference as the cache map value instead of using MapKeyGetter

Co-authored-by: 2No2Name <2no2name@web.de>

2No2Name commented on src/main/java/me/jellysquid/mods/hydrogen/common/state/property/WeakPropertyCache.java:
```diff
@@ -0,0 +1,26 @@
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends Property<?>> T tryCacheProperty(T newProperty) {
+        return (T) CACHE.computeIfAbsent(newProperty, WeakReference::new).get();
```
The WeakReference::get() can probably return null if it is collected after computeIfAbsent returns but before get() is executed. I suggest a local variable and in case of null, put the new property into the map.

magneticflux- replied:
The weak reference just created with the constructor reference is based on `newProperty`, so it won't be collected until `tryCacheProperty` returns. _However_ if `computeIfAbsent` finds a property already in `CACHE` (that isn't `newProperty`), GC could run between `computeIfAbsent` and `get` and finalize+clear the reference which would cause issues.

Do you think something like this would work?
```java
T cachedProperty = (T) CACHE.computeIfAbsent(newProperty, WeakReference::new).get();
if (cachedProperty == null) {
    CACHE.put(newProperty, new WeakReference<>(newProperty));
    return newProperty;
} else {
    return cachedProperty;
}
```
It kinda smells like double-checked locking, but I think it should work because `newProperty` is hard-referenced throughout the entire method.

2No2Name replied:
If that whole thing is still inside the synchronized(CACHE) block, yes I think that will work.

magneticflux- committed:
Make and check a defensive copy of the cached property

This prevents the GC from clearing the weak reference before we access it strongly

magneticflux- committed:
Re-add synchronized block on CACHE

NebelNidas commented:
Why hasn't this been merged yet? Is it still not safe to use?

malte0811 commented:
This doesn't explain why it isn't merged yet (since I'm not the one who would need to merge it), but I'm not sure whether this is the right approach. Nothing is stopping modders from adding their own property implementations, which would not be interned by this PR and continue to cause issues.

Maybe I just don't know Yarn names, but I would have expected one of the mixins to target the property implementation for generic `enum`s, rather than the `Direction`-specific one?

magneticflux- commented:
@malte0811 That's true, if a mod implements `Property<T>` on its own, it won't be cached. That isn't a big issue though because that mod implementation typically won't be used by other mods.

You're right about the mixin, I need to make one that targets the `EnumProperty` explicitly. Static methods can't be overridden, so the `DirectionProperty` mixin is still needed.

magneticflux- committed:
Add missing EnumProperty mixin

malte0811 reviewed (approved):
As discussed on Discord: Mods adding their own property implementation is still a concern, but a purely theoretical one for the time being. So IMO it is reasonable to merge this and deal with the issue if and when it actually turns up (never, likely as not).
